README.md
# rodney_missions
ROS node to control the different missions conducted by the house bot Rodney. It also controls some of the simple jobs.

## Running the Node

Once you have the node built you can run the rodney robot by launching the rodney.launch file.

## Node Information
Topics:
* `missions/mission_request`:  
  Subscribes using MonitorState `std_msgs/String` A request to carry out a robot mission or simple job
* `missions/mission_cancel`:  
  Subscribes `std_msgs/String` A request to cancel the current robot mission
* `missions/acknowledge`:    
  Subscribes `sts_msgs/Empty` An acknowledgement by a user that the mission can move on
* `speech/to_speak`:  
  Publishes `speech/voice` A message with either text containing the words to speak or wav containing the path of the wav file to play
* `robot_face/text_out`:  
  Publishes `std_msgs/String` Text for the robot face to animate the lips to
* `missions/mission_complete`:  
  Publishes `std_msgs/String` Message indicating that the current mission completed
* `commands/lidar_enable`:  
  Publishes `std_msgs/Bool` Message indicating if the LIDAR should be enabled or not

Action Clients:
* `head_control_node`: To request movement of the head
* `face_recognition`: To request and obtain the result of a face recognition scan on an image from the camera
* `move_base`: To reuest that the base is navigated to a new location

Service Clients:
* `stop_motor`: To disable the LIDAR
* `start_motor`: To enable the LIDAR

Parameters:

* `/servo/index0/pan/min`: minimum range for the head pan servo. Default value = -(pi/2.0)
* `/servo/index0/pan/max`: Maximum range for the head pan servo. Default value = pi/2.0
* `/servo/index0/tilt/min`: Minimum range for the head tilt servo. Default value = -(pi/2.0)
* `/servo/index0/tilt/max`: Maximum range for the head tilt servo. Default value = pi/2.0
* `/head/position/pan`: The pan position that the head will move to when a task is complete (default positon). Default value = 0.0
* `/head/position/tilt`: The tilt position that the head will move to when a task is complete (default positon). Default value = 0.0
* `~head/view_step/pan`: The increment that the pan servo will move for each scan. Default value = 0.436332
* `~head/view_step/tilt`: The increment that the tile servo will move for each scan. Default value = 0.436332
* `~head/manual_view_step/pan`: Pan step value to move the camera in manual mode. Default = 0.174533
* `~head/manual_view_step/tilt`: Tilt step value to move the camera in manual mode. Default = 0.174533
* `~head/user_position/pan`: Pan position that the head will move to for user input. Default value = 0.0
* `~head/user_position/tilt`: Tilt position that the head will move to for user input. Default value = -0.5

State Machines:

![alt text](https://github.com/phopley/rodney_missions/blob/master/smach.png "Root State Machine")

The top level state machine contains the following states
* `WAITING`:  
  Is a monitor state for the /missions/mission_request topic
* `PREPARE`:  
  A state that examines a request and either actions a job or hands control to a mission sub state machine
* `REPORT`:  
  A state that reports that a mission is complete
* `DEFAULT_HEAD_POSITION`:  
  A SimpleActionState that returns the head/camera to the default position
* `MOVE_HEAD`:  
  A SimpleActionState that request sthe head/camera be moved

![alt text](https://github.com/phopley/rodney_missions/blob/master/smach_mission1.png "Mission 1 State Machine")

* `MISSION1`:  
  A sub state machine responsible for controlling mission 1, "take a message to.."
  * `PREPARE_MISSION`:  
    A state that loads the given waypoint file
  * `PREPARE_FOR_HEAD_MOVE`:  
    A state which sets up the next required position of the head/camera
  * `MOVE_HEAD`:  
    A SimpleActionState that requests the head/camera be moved 
  * `SCAN_FOR_FACES`:  
    A SimpleActionState that requests the faces recognition operation be carried out on the current image from the camera   
  * `CHECK_FOR_SUBJECT`:  
    A state that checks subjects seen by the camera to the ID of the subject we are searching for. If the subject is seen the message is delivered
  * `WAIT_FOR_USER`:  
    State that waits for the acknowledge message or timesout after 60 seconds
  * `DEFAULT_HEAD_POISTION`:  
    A SimpleActionState that returns the head/camera to the default position
  * `PREPARE_TO_MOVE`:  
    A state machine that constructs a nav goal for the MOVE state
  * `MOVE`:  
    A SimpleActionState that moves the robot base to the given nav goal

![alt text](https://github.com/phopley/rodney_missions/blob/master/smach_mission2.png "Mission 2 State Machine")

* `MISSION2`:  
  A sub state machine responsible for controlling mission 2, "greet who you recognise"
  * `PREPARE_FOR_MOVEMENT_GRT`:  
    A state which sets up the next required position of the head/camera
  * `MOVE_HEAD_GRT`:  
    A SimpleActionState that requests the head/camera be moved
  * `SCAN_FOR_FACES`:  
    A SimpleActionState that requests the faces recognition operation be carried out on the current image from the camera
  * `GREETING`:  
    A state responsible for forming the response to any faces recognised

![alt text](https://github.com/phopley/rodney_missions/blob/master/smach_mission4.png "Mission 4 State Machine")

* `MISSION4`:
  A sub state machine responsible for controlling mission 4, "go home"
  * `PREPARE_MISSION`:  
    A state that loads a given waypoint file and extracts the home location and position
  * `DEFAULT_HEAD_POSITION`:  
    A SimpleActionState that returns the head/camera to the default position
  * `MOVE`:  
    A SimpleActionState that moves the robot base to the given nav goal (home position)

## License
Software source and object files are licensed under the Apache License, Version 2.0. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rodney_missions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Unreleased
------------------
* Mission 4 updated to go to other waypoints not just 'home'

2.0.0 (2019/05/02)
------------------
* Some of the parameters on the parameter server are now private
* Added user input position for screen
* Added Mission 1 "Take a message to ..."
* added Mission 4 "Go Home"

1.1.0 (2019/04/10)
------------------
* Added functionality to enable/disable LIDAR

1.0.1 (2019/01/14)
------------------
* When manual head movement will exceed min/max instead of not moving the head will move to the min/max

1.0.0 (2019/01/12)
------------------
* Head movement now controlled by an action state
* Changed to use radians and the standard ROS orientation
* Changed to use action state to start face recognition
* Sub state machines moved into libraries

0.2.0 (2018-12-05)
------------------
* Added mission 3
* Mission 2 no longer uses action messages

0.1.0 (2018-11-12)
------------------
* First formal release of the package


src/missions_lib/object_search.py
#!/usr/bin/env python
# States for searching for an object

import rospy
from std_msgs.msg import Empty
from smach import State, StateMachine
from smach_ros import SimpleActionState, MonitorState
from actionlib_msgs.msg import GoalStatus
from head_control.msg import point_headAction, point_headGoal
from tf_object_detection.msg import scan_for_objectsAction

# Child (derived) class. Parent class is StateMachine
class Mission3StateMachine(StateMachine):
    def __init__(self, helper_obj):
        StateMachine.__init__(self, outcomes=['complete', 'preempted','aborted'], input_keys=['mission_data','start'])

        self.__helper_obj = helper_obj

        with self:
            # This state will calculate the next head/camera position
            StateMachine.add('PREPARE_FOR_MOVEMENT_OS',
                             PrepareMovementObjectSearch(self.__helper_obj),
                             transitions={'complete':'complete','move':'MOVE_HEAD_OS'},
                             remapping={'start_in':'start','start_out':'start'})
                                 
            # This state will call the action to move the head/camera
            StateMachine.add('MOVE_HEAD_OS',
                             SimpleActionState('head_control_node',
                             point_headAction,                                                            
                             goal_slots=['absolute','pan','tilt']),
                             transitions={'succeeded':'SCAN_FOR_OBJECTS','preempted':'preempted','aborted':'aborted'},
                             remapping={'absolute':'user_data_absolute','pan':'user_data_pan','tilt':'user_data_tilt'}) 
                
            # This state will call the action to scan for objects on the image from the camera.
            StateMachine.add('SCAN_FOR_OBJECTS',
                             SimpleActionState('object_detection',
                             scan_for_objectsAction,
                             result_cb=self.object_detection_result_cb,
                             input_keys = ['mission_data'],
                             output_keys = ['object_detected']),
                             transitions={'succeeded':'CHOICE_OS','preempted':'preempted','aborted':'aborted'})                
                
            # This state will will make the decision which state to move to
            # next. If the object in question was detected then it will move
            # to a state that waits for user interaction. If the object was
            # not detected then it will move to the state to move the camera.
            # This seems a bit of an overkill but the SimpleActionState before it
            # can only provide the three possible outcomes 'succeeded', 'aborted'
            # or 'preempted'. This state will request speech to state the
            # object was found.
            StateMachine.add('CHOICE_OS',
                              ChoiceObjectSearch(self.__helper_obj),           
                              transitions= {'move_camera':'PREPARE_FOR_MOVEMENT_OS',
                                            'await_user':'WAIT_FOR_USER_OS'})

            # This state will wait for a message indicating the user wishes
            # to move on. It will allow the user to examine the camera feed
            # once an object was detected
            StateMachine.add('WAIT_FOR_USER_OS',
                             MonitorState('/missions/acknowledge',
                             Empty,
                             self.UserAckOS),
                             transitions={'valid':'WAIT_FOR_USER_OS',
                                          'invalid':'PREPARE_FOR_MOVEMENT_OS',
                                          'preempted':'preempted'})

    def object_detection_result_cb(self, userdata, status, result):
        if status == GoalStatus.SUCCEEDED:
            # Extract the result results
            if len(result.names_detected) == 0:
                # No objects recognised
                userdata.object_detected = False
            else:
                # Objects detected, but did we find the one we are interested in?
                userdata.object_detected = False
                for n in result.names_detected:
                    if n == userdata.mission_data:
                        userdata.object_detected = True                      
                        break

        # By not returning anything here (at this indentation) the state will 
        # return with the corresponding outcome of the action 

    def UserAckOS(self, userdata, msg):
        # User acknowledged so move on        
        
        # Returning False means the state transition will follow the invalid outcome
        return False

# PREPARE_FOR_MOVEMENT_OS State
class PrepareMovementObjectSearch(State):
    def __init__(self, helper_obj):
        State.__init__(self, outcomes=['complete','move'],
                       input_keys=['start_in'],
                       output_keys=['start_out','user_data_absolute','user_data_pan','user_data_tilt'])
        self.__helper_obj = helper_obj                                 
                       
    def execute(self, userdata):
        # Is this a new search?
        if userdata.start_in == True:
            userdata.start_out = False
            scan_complete = False
            # get the camera start position (pan min and tilt min)
            position_request_pan, position_request_tilt = self.__helper_obj.CameraToStartPos()
        else:            
            scan_complete, position_request_pan, position_request_tilt = self.__helper_obj.CameraToNextPos()

        # Set up user data that will be used for goal in next state if not complete
        userdata.user_data_absolute = True
        userdata.user_data_pan = position_request_pan
        userdata.user_data_tilt = position_request_tilt

        if scan_complete == True:
            greeting = "Mission complete Rodney standing by"            
            self.__helper_obj.Speak(greeting, greeting)
            next_outcome = 'complete'
        else:
            next_outcome = 'move'                
        	    
        return next_outcome

# CHOICE_OS State
class ChoiceObjectSearch(State):
    def __init__(self, helper_obj):
        State.__init__(self, outcomes=['move_camera','await_user'],
                       input_keys=['object_detected','mission_data'])
        self.__helper_obj = helper_obj
                       
    def execute(self, userdata):
        # Now make the decission on the next state to transit to based on if
        # the object was detected on the last scan
        if userdata.object_detected == True:
            next_outcome = 'await_user'
            # We will also request that the robot speaks in response to detecting the object
            greeting = userdata.mission_data + " detected"
            rospy.loginfo(greeting)
            self.__helper_obj.Speak(greeting, greeting)
        else:
            next_outcome = 'move_camera'
        
        return next_outcome    



src/rodney_missions_node.py
#!/usr/bin/env python
# Copyright 2019 Philip Hopley
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not 
# use this file except in compliance with the License. You may obtain a  copy
# of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software 
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.
#
# This ROS node is the state machine for the missions that control Rondey Robot
import sys
import rospy
import roslib; roslib.load_manifest('std_srvs')
import std_srvs.srv
import missions_lib
from std_msgs.msg import String, Empty, Bool
from smach import State, StateMachine
from smach_ros import MonitorState, SimpleActionState, IntrospectionServer
from actionlib_msgs.msg import GoalStatus
from speech.msg import voice
from head_control.msg import point_headAction, point_headGoal
from math import pi

# Helper class to hold code used by serveral different states
class MissionsHelper():
    def __init__(self):
        self.__speech_pub_ = rospy.Publisher('/speech/to_speak', voice, queue_size=1)
        self.__text_out_pub = rospy.Publisher('/robot_face/text_out', String, queue_size=1)
        
        # Obtain values from the parameter server
        # Minium/Maximum range movment of camera
        self.__pan_min = rospy.get_param("/servo/index0/pan/min", -(pi/2.0))
        self.__pan_max = rospy.get_param("/servo/index0/pan/max", pi/2.0)
        self.__tilt_min = rospy.get_param("/servo/index0/tilt/min", -(pi/2.0))
        self.__tilt_max = rospy.get_param("/servo/index0/tilt/max", pi/2.0);
        # Default position after mission ends
        self.__camera_default_pan_position = rospy.get_param("/head/position/pan", 0.0)
        self.__camera_default_tilt_position = rospy.get_param("/head/position/tilt", 0.0)
        # Using '~private_name' will prefix the parameters with the node name given in launch file
        # Step value to move the camera by when searching
        self.__pan_step_value = rospy.get_param("~head/view_step/pan", 0.436332)
        self.__tilt_step_value = rospy.get_param("~head/view_step/tilt", 0.436332)
        # Step value to move the camera in manual mode
        self.__manual_pan_step_value = rospy.get_param("~head/manual_view_step/pan", 0.174533)
        self.__manual_tilt_step_value = rospy.get_param("~head/manual_view_step/tilt", 0.174533)
        # Position to move to for user input
        self.__user_input_position_pan = rospy.get_param("~head/user_position/pan", 0.0)
        self.__user_input_position_tilt = rospy.get_param("~head/user_position/tilt", -0.5)
        
        # When true and scanning pan angle will increase, otherwise decrease
        self.__increase_pan = True
        
        # Position that will be requested to move the head/camera to
        self.__position_request_pan = self.__camera_default_pan_position   
        self.__position_request_tilt = self.__camera_default_tilt_position

        # RPLidar services to start and stop the motor
        rospy.wait_for_service('stop_motor')
        rospy.wait_for_service('start_motor')
        self.__rplidar_stop_motor_srv = rospy.ServiceProxy('stop_motor', std_srvs.srv.Empty)
        self.__rplidar_start_motor_srv = rospy.ServiceProxy('start_motor', std_srvs.srv.Empty)
        # LIDAR should be running but make sure
        self.LidarEnable()        
 
    def Speak(self, text_to_speak, text_to_display):
        voice_msg = voice()
        voice_msg.text = text_to_speak
        voice_msg.wav = ""
        
        # Publish topic for speech and robot face animation
        self.__speech_pub_.publish(voice_msg)
        self.__text_out_pub.publish(text_to_display)
    
    def Wav(self, wav_file, text_to_display):
        voice_msg = voice()
        voice_msg.text = ""
        voice_msg.wav = wav_file
        
        # Publish
        self.__speech_pub_.publish(voice_msg)
        self.__text_out_pub.publish(text_to_display)

    # Function to enable the RPLidar
    def LidarEnable(self):        
        self.__rplidar_start_motor_srv()
        self.__lidar_on = True
        
    # Function to disable the RPLidar
    def LidarDisable(self):        
        self.__rplidar_stop_motor_srv() 
        self.__lidar_on = False 
        
    # Function to Toggle RPLidar on/off
    def ToggleLidar(self):
        if(self.__lidar_on == True):
            self.LidarDisable()
        else:
            self.LidarEnable()     
    
    # Function to return the camera start position when scanning within head movement range         
    def CameraToStartPos(self):
        # Set the camera position to pan min and tilt min
        self.__position_request_pan = self.__pan_min
        self.__position_request_tilt = self.__tilt_max
        # Set the variable that says which direction the pan is going. Start by incrementing
        self.__increase_pan_ = True
        return self.__position_request_pan, self.__position_request_tilt 

    # Function to keep track of position after action to set to default position
    def CameraAtDefaultPos(self, userdata, status, result):
        if status == GoalStatus.SUCCEEDED:
            self.__position_request_pan = self.__camera_default_pan_position
            self.__position_request_tilt = self.__camera_default_tilt_position
        
    # Function returns camera default position
    def CameraDefaultPos(self):
        return self.__camera_default_pan_position, self.__camera_default_tilt_position           
 
    # Function to return the next position when scanning within the head movement range.
    # Also returns indication if all areas scanned or more left           
    def CameraToNextPos(self):
        all_areas_scanned = False
        # Calculate the next position of the head/camera        
        if self.__increase_pan == True:
            if self.__position_request_pan == self.__pan_max:
                # Last scan was at the edge, move tilt up and then pan the other way
                self.__increase_pan = False
                self.__position_request_tilt -= self.__tilt_step_value
                  
                if self.__position_request_tilt < self.__tilt_min:
                    all_areas_scanned = True
            else:
                self.__position_request_pan += self.__pan_step_value
                    
                if self.__position_request_pan > self.__pan_max:
                    # Moved out of range, put back on max
                    self.__position_request_pan = self.__pan_max
        else:    
            if self.__position_request_pan == self.__pan_min:
                # Last scan was at the edge, move tilt up and then pan the other way
                self.__increase_pan = True
                self.__position_request_tilt -= self.__tilt_step_value
                    
                if self.__position_request_tilt < self.__tilt_min:
                    all_areas_scanned = True
            else:
                self.__position_request_pan -= self.__pan_step_value
                    
                if self.__position_request_pan < self.__pan_min:
                    # Moved out of range, put back on min
      	            self.__position_request_pan = self.__pan_min 
      	            
      	if all_areas_scanned == True:
      	    # Reset camera/head position to default values                
            self.__position_request_pan = self.__camera_default_pan_position
            self.__position_request_tilt = self.__camera_default_tilt_position                      
     
        return all_areas_scanned, self.__position_request_pan, self.__position_request_tilt
        
    def CameraManualMove(self, direction):
        relative_request_pan = 0.0
        relative_request_tilt = 0.0
        # Check for up command
        if 'd' in direction:
            relative_request_tilt = self.__manual_tilt_step_value
            if (self.__position_request_tilt + relative_request_tilt) > self.__tilt_max:
                # Would move out of range so move to the max position                   
	            relative_request_tilt = self.__tilt_max - self.__position_request_tilt
	            self.__position_request_tilt = self.__tilt_max
            else:
                # Keep track
                self.__position_request_tilt += relative_request_tilt
                    
        # Check for down command
        if 'u' in direction:
            relative_request_tilt = -(self.__manual_tilt_step_value)
            if (self.__position_request_tilt + relative_request_tilt) < self.__tilt_min:
                # Would move out of range so move to the min position                 
	            relative_request_tilt = self.__tilt_min - self.__position_request_tilt
	            self.__position_request_tilt = self.__tilt_min
            else:
                # Keep track
                self.__position_request_tilt += relative_request_tilt
            
        # Check for left commnand
        if 'l' in direction:
            relative_request_pan = self.__manual_pan_step_value
            if (self.__position_request_pan + relative_request_pan) > self.__pan_max:
                # Would move out of range so move to the max                   
	            relative_request_pan = self.__pan_max - self.__position_request_pan
	            self.__position_request_pan = self.__pan_max
            else:
                # Keep track
                self.__position_request_pan += relative_request_pan
            
        # Check for right command
        if 'r' in direction:

            relative_request_pan = -(self.__manual_pan_step_value)
            if (self.__position_request_pan + relative_request_pan) < self.__pan_min:
                # Would move out of range so so move to min                   
	            relative_request_pan = self.__pan_min - self.__position_request_pan
	            self.__position_request_pan = self.__pan_min
            else:
                # Keep track
                self.__position_request_pan += relative_request_pan            
            
        return relative_request_pan, relative_request_tilt
        
    def UserInputPosition(self):
        return self.__user_input_position_pan, self.__user_input_position_tilt               
    
                          
# The PREPARE state
class Prepare(State):
    def __init__(self, helper_obj):
        State.__init__(self, outcomes=['mission1','mission2','mission4','done_task','head_default','move_head'], 
                       input_keys=['mission'],
                       output_keys=['mission_data','start','user_data_absolute','user_data_pan','user_data_tilt'])
        self.__helper_obj = helper_obj
    
    def execute(self, userdata):        
        # Based on the userdata either change state to the required mission or 
        # carry out single job userdata.mission contains the mission or single 
        # job and a number of parameters seperated by '^'
        retVal = 'done_task';
        
        # Split into parameters using '^' as the delimiter
        parameters = userdata.mission.split("^")
        
        if parameters[0] == 'M1':
            # Mission 1 to search for a known person and deliver a message
            # parameter[1] contains filename of file containing poses for the patrol,
            # the id of the person to search for and the message to deliver            
            userdata.mission_data = parameters[1]
            retVal = 'mission1'        
        elif parameters[0] == 'M2':
            # Mission 2 is scan for faces and greet those known, there are no
            # other parameters with this mission request
            userdata.start = True
            retVal = 'mission2'
        elif parameters[0] == 'M4':            
            # Mission 4 is go home. parameter[1] contains filename of file containing poses
            # It may also contain a waypoint to got to. If not we will go
            userdata.mission_data = parameters[1]            
            retVal = 'mission4'
        elif parameters[0] == 'J1':
            # Simple Job 1 is play a supplied wav file and move the face lips           
            # Publish topic for speech wav and robot face animation
            self.__helper_obj.Wav(parameters[1], parameters[2])
        elif parameters[0] == 'J2':
            # Simple Job 2 is to speak the supplied text and move the face lips
            # Publish topic for speech and robot face animation
            self.__helper_obj.Speak(parameters[1], parameters[2])
        elif parameters[0] == 'J3':
            # Simple Job 3 is to move the head/camera. This command will only
            # be sent in manual mode.
            # parameters[1] will either be 'u', 'd', 'c', 'i' or '-'
            # parameters[2] will either be 'l', 'r' or '-'
            # Check for command
            if 'c' in parameters[1]:
                # Move to default position
                retVal = 'head_default'
            elif 'i' in parameters[1]:
                # Move to user input position. This position is a good position for user input at the screen
                pan_position, tilt_position = self.__helper_obj.UserInputPosition()
                userdata.user_data_absolute = True # This will be a actual position move
                userdata.user_data_pan = pan_position
                userdata.user_data_tilt = tilt_position
                retVal = 'move_head'
            else:                
                relative_request_pan, relative_request_tilt = self.__helper_obj.CameraManualMove(parameters[1]+parameters[2])                
                # Set up user data that will be used for goal in next state
                userdata.user_data_absolute = False # This will be a relative move
                userdata.user_data_pan = relative_request_pan
                userdata.user_data_tilt = relative_request_tilt
                retVal = 'move_head'                
        elif parameters[0] == 'J4':
            # Simple job to toggle the LIDAR on/off
            self.__helper_obj.ToggleLidar()

        return retVal
        

# The REPORT state
class Report(State):
    def __init__(self):
        State.__init__(self, outcomes=['success'], input_keys=['mission_status'])
        self.__pub = rospy.Publisher('/missions/mission_complete', String, queue_size=5)
    
    def execute(self, userdata):        
        # Publishes message that mission completed
        message = userdata.mission_status
        self.__pub.publish(message)
        return 'success'        
        
# Top level state machine. The work for each mission is another state machine
# in the 'mission' states        
class RodneyMissionsNode:
    def __init__(self):
        rospy.on_shutdown(self.ShutdownCallback)
        
        # Subscribe to message to cancel missions        
        self.__cancel_sub = rospy.Subscriber('/missions/mission_cancel', Empty, 
                                             self.CancelCallback)        
        
        # Create an instance of the missions helper class
        self.__missions_helper = MissionsHelper()
        
        # ------------------------- Top level state machine -------------------------
        # Create top level state machine
        self.__sm = StateMachine(outcomes=['preempted','aborted'],                                 
                                 output_keys=['mission_data'])
        with self.__sm:                                   
            # Add a state which monitors for a mission to run            
            StateMachine.add('WAITING',
                             MonitorState('/missions/mission_request',
                             String,
                             self.MissionRequestCB,                             
                             output_keys = ['mission']),
                             transitions={'valid':'WAITING', 'invalid':'PREPARE', 
                                          'preempted':'preempted'})
                             
            # Add state to prepare the mission            
            StateMachine.add('PREPARE',
                             Prepare(self.__missions_helper),                             
                             transitions={'mission1':'MISSION1','mission2':'MISSION2','mission4':'MISSION4',
                                          'done_task':'WAITING','head_default':'DEFAULT_HEAD_POSITION',
                                          'move_head':'MOVE_HEAD'})
                             
            # Add the reporting state
            StateMachine.add('REPORT',
                             Report(),
                             transitions={'success':'DEFAULT_HEAD_POSITION'})

            # Set up action goal for deafult head position            
            default_position_pan, default_position_tilt = self.__missions_helper.CameraDefaultPos()
            head_goal = point_headGoal()
            head_goal.absolute = True
            head_goal.pan = default_position_pan
            head_goal.tilt = default_position_tilt

            # Add the default camera position state. Which moves the head to the default position
            StateMachine.add('DEFAULT_HEAD_POSITION',
                             SimpleActionState('head_control_node',
                                               point_headAction,
                                               result_cb = self.__missions_helper.CameraAtDefaultPos,
                                               goal = head_goal),
                             transitions={'succeeded':'WAITING','preempted':'WAITING','aborted':'aborted'})

            # Add the move head state
            StateMachine.add('MOVE_HEAD',
                             SimpleActionState('head_control_node',
                                               point_headAction,
                                               goal_slots=['absolute', 'pan', 'tilt']),
                             transitions={'succeeded':'WAITING', 'preempted':'WAITING', 'aborted':'aborted'},
                             remapping={'absolute':'user_data_absolute', 'pan':'user_data_pan', 'tilt':'user_data_tilt'}) 

            # ------------------------- Sub State machine for mission 1 ---------------------
            # Create a sub state machine for mission 1 - take a message to
            self.__sm_mission1 = missions_lib.Mission1StateMachine(self.__missions_helper)
            
            # Now add the sub state machine for mission 1 to the top level one
            StateMachine.add('MISSION1', 
                             self.__sm_mission1, 
                             transitions={'complete':'REPORT','preempted':'REPORT','aborted':'REPORT'}) 
            
            # ------------------------- Sub State machine for mission 2 ---------------------
            # Create a sub state machine for mission 2 - face detection and greeting
            self.__sm_mission2 = missions_lib.Mission2StateMachine(self.__missions_helper)

            # Now add the sub state machine for mission 2 to the top level one
            StateMachine.add('MISSION2', 
                             self.__sm_mission2, 
                             transitions={'complete':'REPORT','preempted':'REPORT','aborted':'aborted'}) 
                                                                    
            # ------------------------- Sub State machine for mission 4 ---------------------
            # Create a sub state machine for mission 4 - Go Home
            self.__sm_mission4 = missions_lib.Mission4StateMachine(self.__missions_helper)
            
            # Now add the sub state machine for mission 4 to the top level one
            StateMachine.add('MISSION4', 
                             self.__sm_mission4, 
                             transitions={'complete':'REPORT','preempted':'REPORT','aborted':'REPORT'})            
            # -------------------------------------------------------------------------------
            
            
        # Create and start the introspective server so that we can use smach_viewer
        sis = IntrospectionServer('server_name', self.__sm, '/SM_ROOT')
        sis.start()
                             
        self.__sm.execute()
        
        # Wait for ctrl-c to stop application
        rospy.spin()
        sis.stop()
        
    
    # Monitor State takes /missions/mission_request topic and passes the mission
    # in user_data to the PREPARE state
    def MissionRequestCB(self, userdata, msg):                
        # Take the message data and send it to the next state in the userdata
        userdata.mission = msg.data       
                        
        # Returning False means the state transition will follow the invalid line
        return False
        
    # Callback for cancel mission message
    def CancelCallback(self, data):
        # If a sub statemachine for a mission is running then request it be preempted
        if self.__sm_mission1.is_running():
            self.__sm_mission1.request_preempt() 
        elif self.__sm_mission2.is_running():
            self.__sm_mission2.request_preempt() 
        elif self.__sm_mission4.is_running():
            self.__sm_mission4.request_preempt()                   
        
    def ShutdownCallback(self):        
        self.__sm.request_preempt()
        # Although we have requested to shutdown the state machine 
        # it will not happen if we are in WAITING until a message arrives
        
def main(args):
    rospy.init_node('rodney_missions_node', anonymous=False)
    rospy.loginfo("Rodney missions node started")
    rmn = RodneyMissionsNode()        

if __name__ == '__main__':
    main(sys.argv)

